base: introduce SQLInstanceID (and container)

tbg opened the pull request:
With the advent of multi-tenancy, SQL servers can optionally be
decoupled from the (and share a) KV backend. As a consequence, the
NodeID may not be available and its use across all components that need
to function in a multi-tenant setup is deprecated. Yet, some components
require a similar unique identifier for the SQL server they're a part
of.

This new identifier is the SQLInstanceID. At the time of writing,
since we're not yet able to start true SQL tenant servers and only
exercise parts of this functionality in unit testing, it is a
randomized integer. It's possible that we'll replace it with
a per-tenant counter in production, putting it truly on par with
the NodeID, to be able to give the simple guarantee that these IDs
are never reused between different incarnations of the SQL process
for a given tenant. (Note that initially, at most one SQL tenant
server will be active per tenant at any given time).

This commit introduces an IDContainer which will aid this process by
wrapping both the NodeID and SQLInstanceID, very similar to the work
carried out for Gossip in #47972. We do not yet audit all of the call
sites to keep the initial diff container. Instead, IDContainer has a
Get() method that makes it a drop-in replacement for the previously used
NodeIDContainer. `Get ` will be removed in a future commit while moving
all callers to either the `Optional{,Err}` or `DeprecatedNodeID`
methods.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48110)
<!-- Reviewable:end -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> are you planning on adding another commit to add the `Optional` calls or will that be in another PR?

> Reviewed 17 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48110#-:-M65Sj0W8curKIKWZAZo:b-j5ohbb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

tbg commented:
TFTR! I'll do that here, I just wanted to get a first round of review before I make code changes that could potentially rot. Will ping again when it's ready.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 17 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48110#-:-M65W3sk9115-g_puZxY:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/base/node_id.go, line 88 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48110#-M65VPydE-jD5ON23Ogu:-M65VPydE-jD5ON23Ogv:bfxebre) ([raw file](https://github.com/cockroachdb/cockroach/blob/3275dc82743d20327bb7e1573322ede8c219d65e/pkg/base/node_id.go#L88)):*
> ```Go
> type SQLInstanceID int64
> 
> // IDContainer wraps a SQLInstanceID and optionally a NodeID.
> ```

Should this be called `SQLIDContainer?` The key-value layer should never need one of these, so it's not a direct replacement for `NodeIDContainer`.
___
*[pkg/base/node_id.go, line 116 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/48110#-M65UX8-5sEfWGVox3hV:-M65UX8-5sEfWGVox3hW:bmlojou) ([raw file](https://github.com/cockroachdb/cockroach/blob/3275dc82743d20327bb7e1573322ede8c219d65e/pkg/base/node_id.go#L116)):*
> ```Go
> 		fmt.Fprintf(&buf, "@n%s,", n)
> 	}
> 	return ""
> ```

I don't think this is what you meant.


<!-- Sent from Reviewable.io -->

tbg committed:
base: introduce SQLInstanceID (and container)

With the advent of multi-tenancy, SQL servers can optionally be
decoupled from the (and share a) KV backend. As a consequence, the
NodeID may not be available and its use across all components that need
to function in a multi-tenant setup is deprecated. Yet, some components
require a similar unique identifier for the SQL server they're a part
of.

This new identifier is the SQLInstanceID. At the time of writing, since
we're not yet able to start true SQL tenant servers and only exercise
parts of this functionality in unit testing, it is a randomized integer.
It's possible that we'll replace it with a per-tenant or even global
counter in production, putting it truly on par with the NodeID, to be
able to give the simple guarantee that these IDs are never reused
between different incarnations of the SQL process for a given tenant.
This final decision is kicked down the road, see #48009.

This commit introduces an IDContainer which will aid this process by
wrapping both the NodeID and SQLInstanceID, very similar to the work
carried out for Gossip in #47972. We do not yet audit all of the call
sites to keep the initial diff container. Instead, IDContainer has a
Get() method that makes it a drop-in replacement for the previously used
NodeIDContainer. `Get ` will be removed in a future commit while moving
all callers to either the `Optional{,Err}` or `DeprecatedNodeID`
methods.

Release note: None

tbg committed:
sql: fix up some callers of SQLInstanceID.Get()

Release note: None

tbg committed:
kv: use NodeID only if available

Release note: None

tbg committed:
sql: use SQLInstanceID for leases

Release note: None

tbg committed:
sql: make more NodeID use optional

Release note: None

tbg committed:
server: use trivial DistSQL NodeDescriptor for tenants

Release note: None

tbg committed:
sql: use SQLInstanceID in SchemaChanger

This is only used to log to the (per-tenant) event log.

Release note: None

tbg committed:
sql: change SQLInstanceID to 32 bit

This allows its use for the cluster(= tenant's logical cluster)-wide
unique IDs generator.

Release note: None

tbg committed:
distsql: replace more uses of NodeID.Get()

Release note: None

tbg committed:
backupccl: fail without NodeID

Touches #47970.

Release note: None

tbg committed:
sql: make nodeID optional in transitionCtx

It's used to initialize a txn, where we want to pass zero if we don't
have a NodeID.

Release note: None

tbg committed:
sql: plumb SQLIDContainer further

Release note: None

tbg commented:
bors r=asubiotto,nvanbenschoten

I packed on a lot of commits that remove uses of `Get()`. Will merge so that it doesn't rot over the weekend, but I'll send another PR with any additional review comments (also, I'll remove `Get()`).
